language/move-prover/bytecode/src/borrow_analysis_v2.rs

```diff
@@ -0,0 +1,679 @@
+    function_target::{FunctionData, FunctionTarget},
+    function_target_pipeline::{FunctionTargetProcessor, FunctionTargetsHolder},
+    livevar_analysis::LiveVarAnnotation,
+    stackless_bytecode::{AssignKind, BorrowNode, Bytecode, Operation},
+    stackless_control_flow_graph::StacklessControlFlowGraph,
+};
+use itertools::Itertools;
+use move_model::{
+    ast::TempIndex,
+    model::{FunctionEnv, QualifiedId},
+};
+use std::collections::{BTreeMap, BTreeSet};
+use vm::file_format::CodeOffset;
+
+/// This enum denotes strong edges in the borrow graph
+#[derive(Debug, Clone, Eq, Ord, PartialEq, PartialOrd)]
+pub enum StrongEdge  {
+    Empty,
+    Offset(usize)
+}
+
+impl std::fmt::Display for StrongEdge {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            StrongEdge::Empty => write!(f, "E"),
+            StrongEdge::Offset(offset) => write!(f, "{}", offset),
+        }
+    }
+}
+
+pub enum BorrowEdge {
+    Weak,
+    Strong(StrongEdge)
+}
+
+/// This enum denotes the borrow graph edge domain.
+#[derive(Debug, Clone, Eq, Ord, PartialEq, PartialOrd)]
+pub enum EdgeDomain {
+    Weak,
+    Strong(BTreeSet<StrongEdge>)
+}
+
+impl std::fmt::Display for EdgeDomain {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            EdgeDomain::Weak => write!(f, "*"),
+            EdgeDomain::Strong(edgeset) =>
+                write!(f, "{}", edgeset.iter().map(|edge| format!("{}", edge)).join(", "))
+        }
+    }
+}
+
+impl AbstractDomain for EdgeDomain {
+    fn join (&mut self, other: &Self) -> JoinResult {
+        match (&self, &other) {
+            (Self::Weak, _) => JoinResult::Unchanged,
+            (_, Self::Weak) => {
+                *self = Self::Weak;
+                JoinResult::Changed
+            },
+            (Self::Strong(mut edges1), Self::Strong(edges2)) => {
+                let mut changed = false;
+                for edge in edges2 {
+                    let is_new = edges1.insert(edge.clone());
+                    changed = changed || is_new;
+                }
+                JoinResult::Changed
+            }
+        }
+    }
+}
+
+
+impl EdgeDomain {
+    fn of_singleton (edge: BorrowEdge) -> Self {
+        match edge {
+            BorrowEdge::Weak => EdgeDomain::Weak,
+            BorrowEdge::Strong(s_edge) => {
+                let mut set = BTreeSet::new();
+                set.insert(s_edge);
+                EdgeDomain::Strong(set)
+            }
+        }
+    }
+}
+
+impl std::default::Default for EdgeDomain {
+    fn default() -> Self {Self::Strong(BTreeSet::new())}
+}
+
+
+#[derive(Debug, Clone, PartialEq, PartialOrd, Eq, Default)]
+pub struct BorrowInfo {
+    /// Contains the nodes which are alive. This excludes nodes which are alive because
+    /// other nodes which are alive borrow from them.
+    live_nodes: BTreeSet<BorrowNode>,
```

Comment:
Similar here: It think for all of these, using `SetDomain` instead of `BTreeSet` will lead to simplification in the join.